### PR TITLE
DNS controller and endpoint controller for raven L7

### PR DESCRIPTION
**Pull request by River-sh**
1. DNS manager is used to dynamically update dns records by watching raven agent pod. For instance, the access request from cloud to edge will be forwarded to x-raven-internal-service if a raven-agent is deployed on the destination edge node，it is easy to understand that a request will be sent directly to the destionation edge node if there is no a running raven agent.
2. Service manager is used to maintain the port in x-raven-internal-service, and select a raven agent as the backend for both x-raven-internal-service and x-raven-service in cloud gateway node. You need select a gateway in cloud and label it yurt-app: yurt-tunnel-server, service manager will select automatically the raven agent on the gateway node to undertake the forwarding task.

### CODE REVIEW

**Comment by njucjc** [comment]
Please fix go mod check first.

**Comment by River-sh** [comment]
DNS manager is used to dynamically update dns records by watching raven agent pod. For instance, the access request from cloud to edge will be forwarded to x-raven-internal-service if a raven-agent is deployed on the  destination edge node，it is easy to understand that a request will be sent directly to the destionation edge node if there is no a running raven agent.

**Comment by River-sh** [comment]
Service manager is used to maintain the port in x-raven-internal-service, and select a raven agent as the backend for both x-raven-internal-service and x-raven-service in cloud gateway node. You need select a gateway  in cloud and label it **yurt-app: yurt-tunnel-server**, service manager will select automatically  the raven agent on the gateway node to undertake the forwarding task.

File: config/raven-controller-manager/tunnel/service.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: x-raven-server-svc
```
Comment (rambohe-ch) [inline config/raven-controller-manager/tunnel/service.yaml:4]: x-raven-server-svc --> x-tunnel-server-svc, we should keep consistency with the original svc.

File: config/raven-controller-manager/tunnel/service.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  ports:
+    - name: tcp
+      port: 10262
+      targetPort: 10262
+
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: x-raven-server-internal-svc
```
Comment (rambohe-ch) [inline config/raven-controller-manager/tunnel/service.yaml:22]: x-raven-server-internal-svc --> x-tunnel-server-internal-svc

File: pkg/servicecontroller/worker.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	options := metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", dnscontroller.RavenAgentPodLabelKey, dnscontroller.RavenAgentPodLabelValue),
+		FieldSelector: fmt.Sprintf("%s=%s", "spec.nodeName", nodeName),
+	}
+	ravenAgents, err := sc.kubeClient.CoreV1().Pods(RavenProxyResourceNamespace).List(context.TODO(), options)
+	if err != nil || len(ravenAgents.Items) < 1 {
+		klog.Errorf("failed to get gateway raven agent, %w", err)
+		return err
+	}
+	var gatewayRavenAgent *corev1.Pod = nil
```
Comment (DrmagicE) [inline pkg/servicecontroller/worker.go:116]: The assignment `= nil` here is not necessary. I would suggest `var gatewayRavenAgent *corev1.Pod` instead.

**Pull request closed without merging**